Content.Shared/_CM14/Scoping/SharedScopeSystem.cs

```diff
@@ -0,0 +1,193 @@
+            if (_gameTiming.IsFirstTimePredicted && _gameTiming.InPrediction)
+            {
+                _popupSystem.PopupEntity(msgError, args.Performer, args.Performer);
+            }
+
+            return;
+        }
+
+        if (component.IsScoping)
+            StopScoping(uid, component, args.Performer);
+        else
+            StartScoping(uid, component, args.Performer);
+
+        args.Handled = true;
+    }
+
+    private void OnShutdown(EntityUid uid, ScopeComponent component, ComponentShutdown args)
+    {
+        if (component.User != null)
+        {
+            _actionsSystem.RemoveProvidedActions(component.User.Value, uid);
+            StopScopingHelper(uid, component, component.User.Value);
+        }
+    }
+
+    public bool StartScoping(EntityUid item, ScopeComponent component, EntityUid user)
+    {
+        if (component.IsScoping)
+            return false;
+
+        if (TryComp<ScopeUserComponent>(user, out var scopeUserComp))
+        {
+            scopeUserComp.ScopingItem = item;
+            Dirty(user, scopeUserComp);
+        }
+
+        var msgUser = Loc.GetString("cm-action-popup-scoping-user", ("scope", Name(item)));
+
+        _actionsSystem.SetToggled(component.ScopingToggleActionEntity, true);
+        if (_gameTiming.IsFirstTimePredicted && _gameTiming.InPrediction)
+        {
+            _popupSystem.PopupEntity(msgUser, user, user);
+        }
+
+        StartScopingCamera(user, component);
+
+        component.IsScoping = true;
+        component.LastScopedAt = Transform(user).Coordinates;
+        Dirty(item, component);
+
+        return true;
+    }
+
+    public bool StopScoping(EntityUid item, ScopeComponent component, EntityUid user)
+    {
+        if (!component.IsScoping)
+            return false;
+
+        if (TryComp<ScopeUserComponent>(user, out var scopeUserComp))
+        {
+            scopeUserComp.ScopingItem = null;
+            Dirty(user, scopeUserComp);
+        }
+
+        var msgUser = Loc.GetString("cm-action-popup-scoping-stopping-user", ("scope", Name(item)));
+
+        _actionsSystem.SetToggled(component.ScopingToggleActionEntity, false);
+
+        if (_gameTiming.IsFirstTimePredicted && _gameTiming.InPrediction)
+        {
+            _popupSystem.PopupEntity(msgUser, user, user);
+        }
+
+        StopScopingCamera(user, component);
+
+        component.IsScoping = false;
+        Dirty(item, component);
+
+        return true;
+    }
+
+    private void StopScopingHelper(EntityUid uid, ScopeComponent component, EntityUid user)
+    {
+        if (component.IsScoping)
+            StopScoping(uid, component, user);
+
+        if (!TryComp<HandsComponent>(user, out var hands))
+            return;
+
+        var heldItems = _handsSystem.EnumerateHeld(user, hands).ToArray();
+
+        foreach (var heldItem in heldItems)
+        {
+            if (HasComp<ScopeComponent>(heldItem))
+            {
+                return;
+            }
+        }
```

Comment:
This should be able to enumerate the hands directly so that you don't need to allocate a new array.
```suggestion
        foreach (var heldItem in _handsSystem.EnumerateHeld(user, hands))
        {
            if (HasComp<ScopeComponent>(heldItem))
            {
                return;
            }
        }
```

Reply:
Thanks, updated